snmp_base_block.py

```diff
@@ -0,0 +1,118 @@
+
+
+class SNMPException(Exception):
+    pass
+
+
+class SNMPBase(Block):
+
+    """ A base block implementing a SNMP Manager
+
+    Request SNMP GET ops to SNMP Agents for retrieving monitor data and
+    publish them as signals
+    """
+
+    agent_host = StringProperty(title="SNMP Agent Url", default='127.0.0.1')
+    agent_port = IntProperty(title="SNMP Agent Port", default=161)
+    timeout = TimeDeltaProperty(title='SNMP GET Timeout',
+                                default={"seconds": 1})
+    retries = IntProperty(title="SNMP GET Retries", default=5)
+    exclude_existing = BoolProperty(
+        title="Exclude Existing Values", default=False)
+    lookup_names = BoolProperty(title="Look up OId names", default=False)
+    lookup_values = BoolProperty(title="Look up OId values", default=False)
+    oids = ListProperty(OIDProperty, title="List of OID")
+    version = VersionProperty('0.2.0')
+
+    def __init__(self):
+        super().__init__()
+        self._cmdGen = cmdgen.CommandGenerator()
+        self._data = None
+        self._transport = None
+        self._job = None
+
+    def configure(self, context):
+        """ Configure SNMP by creating data and transport for future
+         GET requests
+        """
+        super().configure(context)
+        self._data = self._create_data()
+        self._transport = cmdgen.UdpTransportTarget(
+            (self.agent_host, self.agent_port),
+            timeout=self.timeout.seconds,
+            retries=self.retries)
+
+    def process_signals(self, signals, input_id='default'):
+        for signal in signals:
+            valid_oids = []
+            for oid in self.oids:
+                try:
+                    next_oid = oid.oid(signal)
+                    valid_oids.append(next_oid)
+                except:
+                    self._logger.exception(
+                        "Could not determine OID from {}".format(oid))
+
+            starting_signal = None if self.exclude_existing else signal
+            if valid_oids:
+                self.execute_request(valid_oids, starting_signal)
+
+    def execute_request(self, oids, starting_signal=None):
+        """ Executes SNMP GET request
+        """
+        try:
+            result = self.make_snmp_request(oids)
+        except SNMPStatusException:
+            # TODO: Make this output on status ouptut
+            self._logger.exception("Error status returned")
+        except SNMPException:
+            self._logger.exception("Error returned")
+
+        if starting_signal:
+            self._handle_data(result, starting_signal)
+        else:
+            self._handle_data(result, Signal())
+
+    def _create_data(self):
+        """ Method to be override in inherited classes
+          Returns the data required by cmdGen for executing SNMP GET
+          operations on an Agent
+        """
+        raise NotImplementedError()
+
+    def make_snmp_request(self, oids):
```

Comment:
Being a little picky here, but why not make make_snmp_request _make_snmp_request?